File: app/models/email_confirmation.rb
# typed: strict
# frozen_string_literal: true

class EmailConfirmation < ApplicationRecord
  extend Enumerize

  EVENT_PASSWORD_RESET = :password_reset
  EVENT_SIGN_UP = :sign_up
  EXPIRES_IN = T.let(15.minutes, ActiveSupport::Duration)

  enumerize :event, in: [EVENT_PASSWORD_RESET, EVENT_SIGN_UP]

  scope :active, -> { where(succeeded_at: nil).where("created_at > ?", EXPIRES_IN.ago) }
  scope :succeeded, -> { where.not(succeeded_at: nil) }

  validates :email, presence: true
  validates :code, format: {with: /\A\d{6}\z/}, presence: true

  sig { returns(String) }
  def self.generate_code
    6.times.map { rand(10) }.join
  end

  sig { returns(T::Boolean) }
  def succeeded?
    !succeeded_at.nil?
  end

  sig { returns(T::Boolean) }
  def success!
    update!(succeeded_at: Time.current) unless succeeded?
    true
  end
end

File: go/internal/model/email_confirmation.go
package model

import (
	"time"
)

// EmailConfirmationEvent はメール確認のイベント種別
type EmailConfirmationEvent string

const (
	// EmailConfirmationEventPasswordReset はパスワードリセットイベント
	EmailConfirmationEventPasswordReset EmailConfirmationEvent = "password_reset"
	// EmailConfirmationEventSignUp はサインアップイベント
	EmailConfirmationEventSignUp EmailConfirmationEvent = "sign_up"
	// EmailConfirmationEventEmailUpdate はメールアドレス更新イベント
	EmailConfirmationEventEmailUpdate EmailConfirmationEvent = "email_update"
)

// EmailConfirmation はメール確認のドメインモデル
type EmailConfirmation struct {
	ID          EmailConfirmationID
	Email       string
	Event       EmailConfirmationEvent
	Code        string
	SucceededAt *time.Time
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

// EmailConfirmationExpirationMinutes は確認コードの有効期限（分）
const EmailConfirmationExpirationMinutes = 15

// IsExpired は確認コードが有効期限切れかどうかを返す
func (ec *EmailConfirmation) IsExpired() bool {
	expirationTime := ec.CreatedAt.Add(time.Duration(EmailConfirmationExpirationMinutes) * time.Minute)
	return time.Now().After(expirationTime)
}

// IsSucceeded は確認が成功済みかどうかを返す
func (ec *EmailConfirmation) IsSucceeded() bool {
	return ec.SucceededAt != nil
}
